File: sorts/ShellSort.java
import java.io.*;
import java.util.*;

/*
  Write a solution to the Shell Sort.
*/
class Solution {
    public static void shellSort(int array[]) {
        int inner, outer, temp, interval = array.length / 2;
        while (interval > 0) {
            for (outer = interval; outer < array.length; outer++) {
                temp = array[outer];
                inner = outer;
                while ((inner > interval - 1) && array[inner - interval] >= temp) {
                    array[inner] = array[inner - interval];
                    inner -= interval;
                }
                array[inner] = temp;
            }
            interval = (interval - 1) / 2;
        }
    }

    public static int[] generateArray(int size) {
        int array[] = new int[size];
        for (int i = 0; i < array.length; i++) {
            array[i] = (int) (Math.random() * 90) + 10;
        }
        return array;
    }

    public static void main(String args[]) {
        int array[] = generateArray(20);
        shellSort(array);
        for (int number : array) {
            System.out.print(number + " ");
        }
    }
}

File: sorts/javascript/shell-sort.js
const shellSort = (array) => {
    let result = [...array];
    let outer, inner, temp, interval = Math.floor((result.length - 1) / 2);
    while (interval > 0) {
        for (outer = interval; outer < result.length; outer++) {
            inner = outer;
            temp = result[outer];
            while ((inner > interval - 1) && result[inner - interval] >= temp) {
                result[inner] = result[inner - interval];
                inner -= interval;
            }
            result[inner] = temp;
        }
        interval = (Math.floor(interval - 1)) / 2;
    }
    return result;
};

console.log(shellSort([6, 1, 3, 9, 10, 11, 4, 5]));
